db/blob/blob_source_test.cc

```diff
@@ -1020,6 +1022,299 @@ TEST_F(BlobSourceTest, MultiGetBlobsFromCache) {
+                                  blob_offsets[0], file_size, blob_sizes[0],
+                                  kNoCompression, nullptr /* prefetch_buffer */,
+                                  &values[0], nullptr /* bytes_read */));
+    ASSERT_EQ(values[0], blobs[0]);
+    ASSERT_TRUE(
+        blob_source.TEST_BlobInCache(file_number, file_size, blob_offsets[0]));
+
+    // key0 should be demoted to the secondary cache, and key1 should be filled
+    // to the primary cache from the blob file.
+    ASSERT_OK(blob_source.GetBlob(read_options, keys[1], file_number,
+                                  blob_offsets[1], file_size, blob_sizes[1],
+                                  kNoCompression, nullptr /* prefetch_buffer */,
+                                  &values[1], nullptr /* bytes_read */));
+    ASSERT_EQ(values[1], blobs[1]);
+    ASSERT_TRUE(
+        blob_source.TEST_BlobInCache(file_number, file_size, blob_offsets[1]));
+
+    OffsetableCacheKey base_cache_key(db_id_, db_session_id_, file_number,
+                                      file_size);
+
+    // blob_cache here only looks at the primary cache since we didn't provide
+    // the cache item helper for the secondary cache. However, since key0 is
+    // demoted to the secondary cache, we shouldn't be able to find it in the
+    // primary cache.
+    {
+      CacheKey cache_key = base_cache_key.WithOffset(blob_offsets[0]);
+      const Slice key0 = cache_key.AsSlice();
+      auto handle0 = blob_cache->Lookup(key0, statistics);
+      ASSERT_EQ(handle0, nullptr);
+      blob_cache->Release(handle0);
+
+      bool found = false;
+      auto sec_handle0 = secondary_cache->Lookup(key0, create_cb, true, found);
+      ASSERT_TRUE(found);
+      ASSERT_NE(sec_handle0, nullptr);
+      ASSERT_TRUE(sec_handle0->IsReady());
+      auto value = static_cast<std::string*>(sec_handle0->Value());
+      ASSERT_EQ(*value, blobs[0]);
+      delete value;
+
+      // For blob source interface, after a cache miss occurs in the primary
+      // cache, key0 can be retrieved in the secondary cache.
+      ASSERT_TRUE(blob_source.TEST_BlobInCache(file_number, file_size,
+                                               blob_offsets[0]));
+    }
+
+    // key1 should exist in the primary cache.
+    {
+      CacheKey cache_key = base_cache_key.WithOffset(blob_offsets[1]);
+      const Slice key1 = cache_key.AsSlice();
+      auto handle1 = blob_cache->Lookup(key1, statistics);
+      ASSERT_NE(handle1, nullptr);
+      blob_cache->Release(handle1);
+
+      bool found = false;
+      auto sec_handle1 = secondary_cache->Lookup(key1, create_cb, true, found);
+      ASSERT_FALSE(found);
+      ASSERT_NE(sec_handle1, nullptr);
+      ASSERT_EQ(sec_handle1->Value(), nullptr);
+
+      ASSERT_TRUE(blob_source.TEST_BlobInCache(file_number, file_size,
+                                               blob_offsets[1]));
+    }
+
+    // key0 promotion should fail due to the blob cache being at capacity,
```

Comment:
I'm assuming it fails because we're not providing the `CacheItemHelper`, so `key1` doesn't get evicted? (If so, this might be worth clarifying)

Reply:
I rewrote the test